### PR TITLE
[TW-1197] Custom functions ES6 and JSON schema support

### PR DESCRIPTION
**Pull request by MatthewPinsker-postman**
* Explained that custom functions now support JavaScript ES6 syntax, in addition to existing support CommonJS syntax
* Updated custom functions examples to prioritize ES6 syntax, but also show CommonJS syntax (commented out)
* Explained that we recommend defining your function options using a JSON schema if your function accepts options
* Added new section about Spectral functions we support (Supported Spectral functions), and added `createRulesetFunction` to the section
* Added section explaining JSON Schema and how to use it with custom functions
* Added a new section "Example: Checking that a value isn't in a list (JSON schema)" explaining how to write a custom function using a JSON schema

[Beta site](https://tw-1197-custom-functions-es6-and-json-schema.learning.postman-beta.com/docs/api-governance/configurable-rules/spectral/)

### CODE REVIEW

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 1]
```diff
@@ -187,7 +192,7 @@ Use the following parameters in your custom functions depending on your use case
 |<div style="width:150px">Parameter</div> | Description
 --- | ---
 `input` | **Required**. This can be any data type, such as a string or array. This is the value that the `given` [JSON Path Plus expression](#json-path-and-json-path-plus) returns. The rule tests the value of `input` using the function.
-`options` | The optional values of `then.functionOptions`. Add this parameter to your function if your function expects options.
+`options` | The optional values of `then.functionOptions`. Add this parameter to your function if your function expects options. <br> If your custom function accepts options, Postman recommends defining the expected options in a [JSON Schema](#json-schema). To learn about adding a JSON Schema to your custom function, see the [`createRulesetFunction` Spectral function](#createrulesetfunction).
```
Comment (mark-anthony-postman) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:195]: Add a second `<br>` so there's some white space between the paragraphs?

**Review by mark-anthony-postman — APPROVED** [review]
LGTM! 🌟

File: src/pages/docs/api-governance/configurable-rules/spectral.md (hunk 1 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/spectral.md:195]: @mark-anthony-postman It doesn't seem to have an effect. Wrapping each paragraph in `<p></p>` worked.

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 2]
```diff
@@ -195,9 +195,9 @@ Use the following parameters in your custom functions depending on your use case
...
+`context` | <p>The third parameter you can add to your function. This optional parameter is used in advanced use cases where you need to investigate several elements. You can use the parameter to access properties about the function. These properties are as follows: </p><ul><li>`path` - A path to an element in the document you want to check for a rule violation. The path must be an array of strings. To learn more, see [Spectral function return statement properties](#spectral-function-return-statement-properties). </li><li>`document` - The document you're attempting to lint.</li><li>`rule` - The rule that's using the function.</li><li>`documentInventory` - Provides access to resolved and unresolved documents, the $ref resolution graph, and other advanced properties.</li></ul>
```
Comment (arno-di-loreto) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:200]: Note that all parameters of the custom function (and so all properties of `context`) are passed by Spectral. Users have no control over them, so they don't decide what's in there. So the descriptions should describe "what is there/what you'll get" and not "what you can put in there."

- `targetVal`:
  - The precision about when `then.field` is present may let users think the value is a JSON Path. It could be rephrased like this: 
    - This is the value that the `given`JSON Path returns. If you defined a value for `then.field` in your rule, it is the value returned by the path composed of the `given` path appended with `then.field`. 
- `context`:
  - "You can use the parameter to access properties about the function." should be rephrased because `context` is not about the function but the context is which it is called:
    - You can use the parameter to access properties about the context in which the function is called.
  - Because users may use `context` to investigate a single element and because there are also other use cases, I think "where you need to investigate several elements" should be removed or replaced by something like
    - ...where, for example, you need to investigate sub-elements of `targetVal` or other locations in the document 
  - `path`:
    - The "a path to ..." and "must be an array" may let users think can mistakenly change that, which is not the case (and the 3 other properties are defined in "this is what you'll get" mode). Also, the connection with `targetVal` is not shown. And the learn more may be rephrased to tell users can learn how to use it in a later section. Overall, it could be rephrased something like
      - The path to `targetVal` as an array of string, `["paths", "/resources", "get", "responses", "306"]`, for example. To learn how to use it, see ...

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 3]
```diff
@@ -195,9 +195,9 @@ Use the following parameters in your custom functions depending on your use case
...
-`context` | <p>The third parameter you can add to your function. This optional parameter is used in advanced use cases where you need to investigate several elements. You can use the parameter to access properties about the function. These properties are as follows: </p><ul><li>`path` - The `given` [JSON Path Plus expression](#json-path-and-json-path-plus) that `targetVal` points to. </li><li>`document` - The document you're attempting to lint.</li><li>`rule` - The rule that's using the function.</li><li>`documentInventory` - Provides access to resolved and unresolved documents, the $ref resolution graph, and other advanced properties.</li></ul>
+`targetVal` | **Required**. The first parameter you must add to your function. This can be any data type, such as a string or array. This is the value that the `given` [JSON Path Plus expression](#json-path-and-json-path-plus) returns. If you also define a value for `then.field` in your rule, `targetVal` returns the `given` path appended with the `then.field` value. The rule tests the value of `targetVal` using your custom function.
```
Comment (arno-di-loreto) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:198]: - `targetVal`:
  - The precision about when `then.field` is present may let users think the value is a JSON Path. It could be rephrased like this: 
    - This is the value that the `given`JSON Path returns. If you defined a value for `then.field` in your rule, it is the value returned by the path composed of the `given` path appended with `then.field`.

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 4]
```diff
@@ -210,7 +210,7 @@ Use the following properties to write the return statement in your custom functi
...
+`path` | <p>An optional path to an element in the document you want to check for a rule violation. The path must be an array of strings. The `path` property will return when investigating sub-elements of the `targetVal` value or other locations in the document. If you use the `path` property, you must also use the [`context` parameter](#spectral-function-parameters) in your function.</p> <p>If you don't add the `path` property, the default path is the `targetVal` value.</p> <p>When targeting sub-elements of `targetVal`, use the following syntax to add the `path` property to your function and append a custom path: `path: ["custom", "path"]`.</p>
```
Comment (arno-di-loreto) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:213]: - `path`
  - Maybe it's my Frenglish, "An optional path to an element you want to check for a rule violation" sounds like we're before the check, but we're working on returning the result of the check. Maybe rephrase to "An path to the element triggering the rule violation"
  -  Maybe add an example after "The path must be an array of strings" `["paths", "/resources", "get", "responses", "306"]` (note: it's the same example I proposed for `context.path`)
  - Again, maybe my Frenglish, "the path property will return when ... " sounds like users have no control over it, which is not the case here. Maybe rephrase "the path property should be returned when ..."
  - Maybe add an explanation of why "must use context"
    - If you use the `path`property, you'll need the [`context` parameter](#spectral-function-parameters) in your function to get access to the full document or the path to `targetVal`.
  - "If you don't add ..." could be moved erlier or at the end so users see one reason why "you must use context" right after it is said (and how to use it). 
  - "the default path is the `targetVal`~~value~~ path" or "path to `targetVal`" (hence it's not the value of `targetVal`)
  -  "When targering sub-elements..."
     - Maybe my Frenglish but you don't "add the path property to your function"; you build the path (to return it) or "add it to the function's return"
     - The sub-element of `targetVal` example is incorrect; the original version was correct `path: [...context.path, "a", "custom", "path"]`
     - Rephrase/fix could be like this:
       - When targeting sub-elements of the `targetVal` value, use the following syntax to build `path`:  `path: [...context.path, "path", "to", "sub-element"]`. The `context.path` contains the path to `targetVal`.

Full edit:

An optional path to the element triggering the rule violation. The path must be an array of strings, `["paths", "/resources", "get", "responses", "306"]`, for example. 

If you don't add the `path` property, the default path is the `targetVal` value.

The `path`property should be returned when investigating sub-elements of the `targetVal` value or other locations in the document. If you use the `path` property, you'll need the [`context` parameter](#spectral-function-parameters) in your function to get access to the full document or the path to `targetVal`. 

When targeting sub-elements of the `targetVal` value, use the following syntax to build `path`:  `path: [...context.path, "path", "to", "sub-element"]`. The `context.path` contains the path to `targetVal`.

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 5]
```diff
@@ -250,24 +250,24 @@ You can import the following Spectral functions into your custom function file i
...
-`options` | **Required**. Postman expects a key named `options` to be second in the JSON object. This is the JSON Schema that defines the values of the [`then.functionOptions` property](#spectral-rule-properties) in your rule, which are your custom function's options. Your custom function tests this using the [`options` parameter](#spectral-function-parameters). This JSON Schema enables you to validate whether the options match specific criteria.
+`input` | <p>**Required**. Postman expects a key named `input` in the JSON object. This is the JSON Schema that defines the value that the [`targetVal` parameter](#spectral-function-parameters) returns in your custom function. This JSON Schema enables you to validate whether the value of `targetVal` matches specific criteria.</p> <p>The `targetVal` value must match the `input` JSON Schema; otherwise, the custom function won't execute.</p> <p>Enter the following to skip validation for the `input` JSON Schema: `input: null,`.</p>
```
Comment (arno-di-loreto) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:263]: - "This is the JSON Schema that defines the value that the [`targetVal` parameter](#spectral-function-parameters) returns in your custom function. "
  - This is the JSON Schema of the [`targetVal` parameter](#spectral-function-parameters) in your custom function.

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 6]
```diff
@@ -250,24 +250,24 @@ You can import the following Spectral functions into your custom function file i
...
+`options` | **Required**. Postman expects a key named `options` in the JSON object. This is the JSON Schema that defines the values that the [`options` parameter](#spectral-function-parameters) returns in your custom function. This JSON Schema enables you to validate whether the value of `options`` match specific criteria.
```
Comment (arno-di-loreto) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:264]: - "This is the JSON Schema that defines the values that the [`options` parameter](#spectral-function-parameters) returns in your custom function."
  - This is the JSON Schema of the [`options` parameter](#spectral-function-parameters) in your custom function.
-  "This JSON Schema enables you to validate whether the value of `options`` match specific criteria."
   - Typo: extra backtick after options 
   - A link to the options validation explanation (JSON Schema section) would be nice
   - That would give:
     - This JSON Schema enables you to validate whether the value of `options` match specific criteria. To learn more, see [JSON Schema](#user-content-json-schema).

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 7]
```diff
@@ -363,9 +359,9 @@ The following examples show JSON Schema key-value pairs you can use to validate
...
+> If your custom function accepts options, Postman recommends adding a [JSON Schema](#json-schema) to your custom function. A JSON Schema enables you to define and validate your custom function's options when editing your rule. This requires you to export your custom function using the [`createRulesetFunction` Spectral function](#createrulesetfunction).
 
-The custom function checks the value of the option `values`, which is defined in the [Spectral document](#example-rule-that-uses-a-custom-function) (or ruleset) using `then.functionOptions`. The value of `values` is a list of numeric strings. If the `targetVal` path returns a value already in the list, the rule violation is triggered.
+The custom function checks the value of the option `values`, which is defined in the [Spectral document](#example-rule-that-uses-a-custom-function) (or ruleset) using `then.functionOptions`. The value of `values` is a list of numeric strings. If `targetVal` returns a value already in the list, the rule violation is triggered.
```
Comment (arno-di-loreto) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:364]: Maybe my Frenglish, "If `targetVal`~~returns~~  a value ...", "If `targetVal`**is**  a value ..."?

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 8]
```diff
@@ -395,9 +391,9 @@ The following custom function named `notInEnumeration` is in a file named `not_i
 
 Before the custom function, the [`createRulesetFunction` Spectral function](#createrulesetfunction) is imported into the file. This enables you to define the expected options in a [JSON Schema](#json-schema) to validate whether the [provided options in your rule](/docs/api-governance/configurable-rules/configuring-api-governance-rules/#adding-custom-rules) match specific criteria.
 
-The custom function checks the value of the option `values`, which is defined in the [Spectral document](#example-rule-that-uses-a-custom-function) (or ruleset) using `then.functionOptions`. The value of `values` is a list of numeric strings. If the `targetVal` path returns a value already in the list, the rule violation is triggered.
+The custom function checks the value of the option `values`, which is defined in the [Spectral document](#example-rule-that-uses-a-custom-function) (or ruleset) using `then.functionOptions`. The value of `values` is a list of numeric strings. If `targetVal` returns a value already in the list, the rule violation is triggered.
```
Comment (arno-di-loreto) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:394]: Same as line 364. 
Maybe my Frenglish, "If targetValreturns a value ...", "If targetValis a value ..."?

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 9]
```diff
@@ -395,9 +391,9 @@ The following custom function named `notInEnumeration` is in a file named `not_i
...
 
-After the custom function, `export default` or `module.exports` calls the `createRulesetFunction` Spectral function and includes the following arguments: a JSON object containing JSON Schemas defining the value that the `given` property returns and the custom function's expected options, and the custom function's name. This exports the custom function so the rule can add it using `then.function`.
+After the custom function, `export default` or `module.exports` calls the `createRulesetFunction` Spectral function and includes the following arguments: a JSON object containing JSON Schemas defining the values of the `targetVal` parameter and `options` parameter, and the custom function's name. This exports the custom function so the rule can add it using `then.function`.
```
Comment (arno-di-loreto) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:396]: Not sure about this, but the phrasing of "a JSON object containing JSON Schemas defining the values of the `targetVal` parameter and `options` parameter" may let users think these are actual value  -> "a JSON object containing the JSON Schemas of the `targetVal` parameter and `options` parameter"

File: src/pages/docs/api-governance/configurable-rules/spectral.md [hunk 10]
```diff
@@ -448,7 +444,7 @@ The following Spectral document has a rule named `http-status-obsolete` that use
 
 The [custom function](#example-checking-that-a-value-isnt-in-a-list-json-schema) is added to the rule using `then.function`. The value of `then.function` is the filename `not_in_enumeration`.
 
-The custom function accepts options using `then.functionOptions` as a property named `values` that's a list of numeric strings. The value of `then.functionOptions.values` is passed to the custom function `notInEnumeration`. The custom function then checks whether a rule violation occurred at the `given` path.
+The custom function accepts options using `then.functionOptions` as a property named `values` that's a list of numeric strings. The value of `then.functionOptions.values` is passed to the custom function `notInEnumeration`. The custom function then checks whether a rule violation occurred at the `given` path appended with the value of `then.field`.
```
Comment (arno-di-loreto) [inline src/pages/docs/api-governance/configurable-rules/spectral.md:447]: - "The custom function then checks whether a rule violation occurred at the `given` path appended with the value of `then.field`."
  - Note: the example shows a specific use-case of `then.field`. This case is described in the "Spectral rule properties" section (in the rule properties list). The `@field` value indicates the keys within the object found by the `given`path. So the function will be called for each key found in the object returned by the `given`path. 
  - So if precisions make sense, we may add something like
    - ... appended with the value of `then.field`. In that case, the `@key` being used, the function will be called for each key in the objects found by the `given` path.
    - Note we could also add more "subject matter" information (the rule checks the presence of obsolete HTTP status codes in the responses object of all operations, the codes are the keys inside the responses object) to help understand how all this works, but we can keep that for the next update.

**Review by arno-di-loreto — APPROVED** [review]
Release specific content is OK (other elements to fix later).

File: src/pages/docs/api-governance/configurable-rules/spectral.md (hunk 2 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/spectral.md]: This is now addressed. Please see https://github.com/postmanlabs/postman-docs/pull/4903/commits/321c203ac7cbbac0f7f0ff94f7d1844b66e2b3d3

File: src/pages/docs/api-governance/configurable-rules/spectral.md (hunk 4 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/spectral.md]: This is now addressed. Please see https://github.com/postmanlabs/postman-docs/pull/4903/commits/321c203ac7cbbac0f7f0ff94f7d1844b66e2b3d3

File: src/pages/docs/api-governance/configurable-rules/spectral.md (hunk 6 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/spectral.md]: This is now addressed. Please see https://github.com/postmanlabs/postman-docs/pull/4903/commits/321c203ac7cbbac0f7f0ff94f7d1844b66e2b3d3

File: src/pages/docs/api-governance/configurable-rules/spectral.md (hunk 7 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/spectral.md]: This is now addressed. Please see https://github.com/postmanlabs/postman-docs/pull/4903/commits/321c203ac7cbbac0f7f0ff94f7d1844b66e2b3d3

File: src/pages/docs/api-governance/configurable-rules/spectral.md (hunk 8 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/spectral.md]: This is now addressed. Please see https://github.com/postmanlabs/postman-docs/pull/4903/commits/321c203ac7cbbac0f7f0ff94f7d1844b66e2b3d3

File: src/pages/docs/api-governance/configurable-rules/spectral.md (hunk 9 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/spectral.md]: This is now addressed. Please see https://github.com/postmanlabs/postman-docs/pull/4903/commits/321c203ac7cbbac0f7f0ff94f7d1844b66e2b3d3

File: src/pages/docs/api-governance/configurable-rules/spectral.md (hunk 10 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/spectral.md]: This will be addressed in https://postmanlabs.atlassian.net/browse/TW-1233 . For now, I added a link to the rule properties section in the docs.

File: src/pages/docs/api-governance/configurable-rules/spectral.md (hunk 3 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/spectral.md]: This is now addressed. Please see https://github.com/postmanlabs/postman-docs/pull/4903/commits/321c203ac7cbbac0f7f0ff94f7d1844b66e2b3d3

**Review by raghavgarg1257 — APPROVED** [review]

**Review by codebyshubham — DISMISSED** [review]

**Review by ashokmadhavi — APPROVED** [review]

File: src/pages/docs/api-governance/configurable-rules/configuring-custom-governance-functions.md [hunk 11]
```diff
@@ -19,7 +19,7 @@ You can create custom governance functions and use them in your [custom API Gove
 
 ## Adding a custom function
 
-You can add custom governance functions to use in your custom API Governance rules. Postman provides you with a boilerplate function to help you start writing your custom governance function. You must write custom functions in JavaScript using CommonJS syntax.
+You can add custom governance functions to use in your custom API Governance rules. Postman provides you with a boilerplate function to help you start writing your custom governance function. You must write custom functions in JavaScript using ES6 syntax and CommonJS syntax.
```
Comment (codebyshubham) [inline src/pages/docs/api-governance/configurable-rules/configuring-custom-governance-functions.md:22]: Will it make more sense to use `or` instead of `and` here since the persona can use either syntax?
```md
ES6 syntax or CommonJS syntax
```

**Review by mark-anthony-postman — APPROVED** [review]
LGTM! 😀

File: src/pages/docs/api-governance/configurable-rules/configuring-custom-governance-functions.md (hunk 11 above)
Comment (MatthewPinsker-postman) [reply src/pages/docs/api-governance/configurable-rules/configuring-custom-governance-functions.md]: This is now fixed. Thanks!

**Review by codebyshubham — APPROVED** [review]

**Review by arno-di-loreto — APPROVED** [review]

**Review by jkonrath-postman — APPROVED** [review]
LGTM
